### PR TITLE
Feature#2081

### PR DESCRIPTION
**Pull request by EdinCitaku**
https://github.com/yairm210/Unciv/issues/2081
Hope I understood the issue correctly :)
I added a movement penalty, so that all newly constructed units start with movement 0f,
and no unit with the same UnitType as already stationed in the city can be constructed.

### CODE REVIEW

**Comment by JackRainy** [comment]
In some cases civilian units (e.g. workers or Great People) are granted for free (e.g. building The Pyramids provides two workers). Will your change affect this?

File: core/src/com/unciv/models/ruleset/unit/BaseUnit.kt [hunk 1]
```diff
@@ -160,6 +160,9 @@ class BaseUnit : INamed, IConstruction {
             && construction.cityInfo.containsBuildingUnique("All newly-trained melee, mounted, and armored units in this city receive the Drill I promotion"))
             unit.promotions.addPromotion("Drill I", isFree = true)
 
+        //movement penalty
+        if(unit.name != "Landsknecht")
+        unit.currentMovement = 0f
```
Comment (r3versi) [inline core/src/com/unciv/models/ruleset/unit/BaseUnit.kt]: Instead of checking by `name`, I think we should add a unique to Landsknecht and check for it here.

**Comment by EdinCitaku** [comment]
> In some cases civilian units (e.g. workers or Great People) are granted for free (e.g. building The Pyramids provides two workers). Will your change affect this?

I dont think so, I just checked the code and those units are not created by construction in the buy screen, as can be seen here:

`        if ("Provides 2 free workers" in uniques) {
            civInfo.placeUnitNearTile(construction.cityInfo.location, Constants.worker)
            civInfo.placeUnitNearTile(construction.cityInfo.location, Constants.worker)
        }
`

File: core/src/com/unciv/models/ruleset/unit/BaseUnit.kt (hunk 1 above)
Comment (EdinCitaku) [reply core/src/com/unciv/models/ruleset/unit/BaseUnit.kt]: ok I addded a new unique for this :)

**Comment by EdinCitaku** [comment]
> Is this `when` operator better than a regular `if` condition?
I dont know if its better per se, but I think its quite readable, esp since when does implicit type casting (maybe if does too?).


> If I already have a missile or an airplane in the city, can I build another one?
No u cant! So I should probably change it? Should u be able to buy multible planes in the same round? or only one plane per round? I am not sure how it is done in normal civ v tbh

**Comment by JackRainy** [comment]
The city has a limit for 6 air units, so you should be able to buy/build them (maybe not the same turn?) even if some of them are already in the city. Also, take into account the carriers with the planes on board which may be docked in the city.

**Comment by EdinCitaku** [comment]
> The city has a limit for 6 air units, so you should be able to buy/build them (maybe not the same turn?) even if some of them are already in the city. Also, take into account the carriers with the planes on board which may be docked in the city.

Ok I just added this functionality :) Just need to get the test to pass haha

File: core/src/com/unciv/ui/cityscreen/ConstructionsTable.kt [hunk 2]
```diff
@@ -300,7 +302,9 @@ class ConstructionsTable(val cityScreen: CityScreen) : Table(CameraStageBaseScre
                         selectedQueueEntry = -2
                         cityScreen.selectedConstruction = null
                     }
-                    if (!construction.shouldBeDisplayed(cityConstructions)) cityScreen.selectedConstruction = null
+                    if (!construction.shouldBeDisplayed(cityConstructions))
+                        cityScreen.selectedConstruction = null
+                    city.boughtPlaneThisRound = true
```
Comment (JackRainy) [inline core/src/com/unciv/ui/cityscreen/ConstructionsTable.kt]: It seems this flag will be set even if I buy not only the airplane - is it expected?

File: core/src/com/unciv/logic/city/CityInfo.kt [hunk 3]
```diff
@@ -582,5 +585,28 @@ class CityInfo {
         for(otherCiv in civsWithCloseCities)
             otherCiv.getDiplomacyManager(civInfo).setFlag(DiplomacyFlags.SettledCitiesNearUs,30)
     }
+
+    private fun hasUnitOfType(type: UnitType) : Boolean
+    {
+        for(unit in getCenterTile().getUnits()) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/city/CityInfo.kt]: This can be simplified to return getUnits().any{it.type==type}

File: core/src/com/unciv/logic/city/CityInfo.kt [hunk 4]
```diff
@@ -582,5 +585,28 @@ class CityInfo {
...
+            if(unit.type == type) true
```
Comment (JackRainy) [inline core/src/com/unciv/logic/city/CityInfo.kt]: Is `return` missing here?

File: core/src/com/unciv/logic/city/CityInfo.kt [hunk 5]
```diff
@@ -52,6 +54,7 @@ class CityInfo {
     var hasSoldBuildingThisTurn = false
     var isPuppet = false
 
+    var boughtPlaneThisRound = false
```
Comment (JackRainy) [inline core/src/com/unciv/logic/city/CityInfo.kt]: It is too specific for the planes. Do we need the same for other types? I would not like to get much of these here.

File: core/src/com/unciv/ui/cityscreen/ConstructionsTable.kt (hunk 2 above)
Comment (EdinCitaku) [reply core/src/com/unciv/ui/cityscreen/ConstructionsTable.kt]: oh good catch lol, dont know how I missed that one!

File: core/src/com/unciv/logic/city/CityInfo.kt (hunk 5 above)
Comment (EdinCitaku) [reply core/src/com/unciv/logic/city/CityInfo.kt]: hmm thats a good question, right now we dont need this for any other type. Alternatively I could make a list of unittypes bought this round?

File: core/src/com/unciv/logic/city/CityInfo.kt (hunk 3 above)
Comment (EdinCitaku) [reply core/src/com/unciv/logic/city/CityInfo.kt]: very true!

File: core/src/com/unciv/models/ruleset/unit/BaseUnit.kt [hunk 6]
```diff
@@ -160,6 +160,9 @@ class BaseUnit : INamed, IConstruction {
             && construction.cityInfo.containsBuildingUnique("All newly-trained melee, mounted, and armored units in this city receive the Drill I promotion"))
             unit.promotions.addPromotion("Drill I", isFree = true)
 
+        //movement penalty
+        if(!unit.hasUnique("Starts with no movement penalty"))
+        unit.currentMovement = 0f
```
Comment (JackRainy) [inline core/src/com/unciv/models/ruleset/unit/BaseUnit.kt]: What if I build a unit, not buy it? Should it also be built with a movement penalty?

File: core/src/com/unciv/logic/city/CityInfo.kt [hunk 7]
```diff
@@ -582,5 +585,27 @@ class CityInfo {
...
+
+    fun canPurchase(construction : IConstruction) : Boolean
+    {
+        if(construction is BaseUnit)
+            if(construction.unitType.isAirUnit()) {
+                val alreadyBoughtPlane = typeOfUnitsBought.any { it.isAirUnit() }
+                return alreadyBoughtPlane && CanStationAdditionalAirUnit()
+            }
+        else
+            return !(hasUnitOfType(construction.unitType))
```
[the remark this answers is not in the document]
Comment (EdinCitaku) [reply core/src/com/unciv/logic/city/CityInfo.kt]: Oh I misremembered, I thought that newly bought units get stacked together, which apparently they wont!

File: core/src/com/unciv/models/ruleset/unit/BaseUnit.kt (hunk 6 above)
Comment (EdinCitaku) [reply core/src/com/unciv/models/ruleset/unit/BaseUnit.kt]: thats a good point, for this I think there is no way than to refactor the postBuildEvent function and give it a parameter like "wasBought: Boolean" with a defaultvalue of false

**Comment by yairm210** [comment]
Looks good to go!
It's conflicting, but once you solve the conflicts I'll merge it =)

**Comment by EdinCitaku** [comment]
>  It's conflicting, but once you solve the conflicts I'll merge it =)

DONT hahahaha, I shouldnt have pushed this. I get some Exception in the canPurchase() function, so I assume there is some problem. Investigating where the issue comes from!
"Exception in thread "LWJGL Application" kotlin.UninitializedPropertyAccessException: lateinit property currentTile has not been initialized"

**Comment by JackRainy** [comment]
Yes, it still crashes with `UninitializedPropertyAccessException`.

I see you are trying to use `canMoveTo` to test whether the unit can reach the city center where it is built. Smart trick! However, that's an overkill.
You do not need a lot of tests which are called within `canMoveTo`.

Let's take a look together:
```
    fun canMoveTo(tile: TileInfo): Boolean {
        if (unit.type.isAirUnit())
            return canAirUnitMoveTo(tile, unit)

/// **The city center is always passable as terrain. Do not need this.**
        if (!canPassThrough(tile))  
            return false

/// **The checks below for owner are obvious. Hmm.. which nation can be in YOUR city? Do not need this.**
        if (unit.type.isCivilian())
            return tile.civilianUnit == null && (tile.militaryUnit == null || tile.militaryUnit!!.owner == unit.owner)  
        else return tile.militaryUnit == null && (tile.civilianUnit == null || tile.civilianUnit!!.owner == unit.owner)
    }
```
Out of this, you need only check for number of air units, and for civilian and military unit presence.

And if you think for one more turn ahead, you will realise that after purchasing there is a call of `postBuildEvent` which calls `placeUnitNearTile` and it already does a test for air units placement!
That's mean the check for number of air units is also not needed!!

Summing up, you need only a check for civilian and military unit presence.

File: android/assets/jsons/Units.json [hunk 8]
```diff
@@ -530,7 +530,7 @@
 		"strength": 16,
 		"cost": 45,
 		"requiredTech": "Civil Service",
-		"uniques": ["Bonus vs Mounted 50%", "Starts with no movement penalty" ],
+		"uniques": ["Bonus vs Mounted 50%", "Can move directly once bought" ],
```
Comment (JackRainy) [inline android/assets/jsons/Units.json]: Why did you decide to change the wording? The initial one seems to be better.
I understand "directly" like "straight". Maybe, "immediately" fits better?

File: android/assets/jsons/translations/English.properties [hunk 9]
```diff
@@ -3187,6 +3187,8 @@ Haka War Dance =
 Rejuvenation = 
  # Requires translation!
 All healing effects doubled = 
+Starts with no movement penalty =
```
Comment (JackRainy) [inline android/assets/jsons/translations/English.properties]: If you change wording, then this line is not needed.

File: core/src/com/unciv/ui/cityscreen/ConstructionsTable.kt [hunk 10]
```diff
@@ -11,6 +11,7 @@ import com.unciv.logic.city.CityInfo
 import com.unciv.logic.city.IConstruction
 import com.unciv.logic.city.SpecialConstruction
 import com.unciv.models.UncivSound
+import com.unciv.models.ruleset.unit.BaseUnit
```
Comment (JackRainy) [inline core/src/com/unciv/ui/cityscreen/ConstructionsTable.kt]: This line seems to be not needed.

File: core/src/com/unciv/logic/city/CityInfo.kt [hunk 11]
```diff
@@ -17,6 +17,8 @@ import com.unciv.logic.trade.TradeOffer
 import com.unciv.logic.trade.TradeType
 import com.unciv.models.ruleset.tile.ResourceSupplyList
 import com.unciv.models.ruleset.tile.ResourceType
+import com.unciv.models.ruleset.unit.BaseUnit
+import com.unciv.models.ruleset.unit.UnitType
```
Comment (JackRainy) [inline core/src/com/unciv/logic/city/CityInfo.kt]: These lines seem to be not needed.

File: android/assets/jsons/Units.json (hunk 8 above)
Comment (EdinCitaku) [reply android/assets/jsons/Units.json]: I changed it cause I felt like the meaning of "movement penalty" is not immediately clear esp if u just read the description. Additionally it was kind of clumsy to translate into german.
